### DISCUSSION TITLE
Add support for updating inverter registers

### DISCUSSION THREAD

**Discussion by kclif9** [squishykid/solax#121] [Ideas]
Some registers are available through the http://5.8.8.8/?optType=ReadSetData API. It would be great if we could have support for modifying these registers through the HA integration / solax library. This would let us adjust settings based on the wholesale energy price, or switch work modes when we're wanting to import from grid and protect the battery charge% during low energy prices. 

Based on some calls with my X1-Hybrid G3, these are the registers I could ascertain the values from:

| Register ID | Register Description | Sample Value |
| ----------- | -------------------- | -------------- |
| 30 | Work Mode | See Notes | 
| 31 | Minimum Capacity (%) | 10 | 
| 35 | Charge Max Current (A) | 20 | 
| 36 | Discharge Max Current (A) | 20 | 
| 37 | Force Time Use Charge Period 1 Start Time (H) | 13 | 
| 38 | Force Time Use Charge Period 1 Start Minute (M) | 5 | 
| 39 | Force Time Use Charge Period 1 End Time (H) | 15 | 
| 40 | Force Time Use Charge Period 1 End Time (M) | 0 | 
| 45 | Force Time Use Charge Period 2 Start Time (H) | 0 | 
| 46 | Force Time Use Charge Period 1 Start Minute (M) | 5 | 
| 47 | Force Time Use Charge Period 1 End Time (H) | 15 | 
| 48 | Force Time Use Charge Period 1 End Time (M) | 0 | 
| 61 | Force Time Use Charge Period Enablement | See notes |
| 135 | Backup Mode Charge From Grid Enabled | See notes |
| 136 | Backup Mode Charge Start Time (H) | 1 |
| 137 | Backup Mode Charge Start Time (M) | 0 |
| 138 | Backup Mode Charge End Time (H) | 2 |
| 139 | Backup Mode Charge End Time (M) | 30 |

Registers 53 - 56 contain the Static IP settings. Registers 57 - 60 contain the Default Gateway settings. Registers 153-156 contain the subnet mask. Although I can't see these getting used. 

Some notes:

1. Work mode for X1-Hybrid: 0 = Self Use, 1 = Force Time Use, 2 = Backup Mode, 3 = Feed In Priority
2. Hours are always in 24 hours - Single digit if under 10 (I.e. 0 to 23)
3. Minutes have single digit if under 10 (I.e. 0 to 59)
4. FTU Charge Period Enablement: 0 = Both Disabled, 1 = Charge Period 1 Enabled, 2 = Charge Period 2 Enabled, 3 = Both Charge Periods Enabled
5. Backup Mode Charge From Grid Enabled: 0 = Disabled, 1 = Enabled

**Comment by squishykid** [comment]
I think this is a great idea! I would feel a bit more confident about implementing it if we had a better overview of which inverters supported this. Do you have some reference material?

**Reply by kclif9** [reply]
I don't have any reference material unfortunately. I saw the API listing over on the solar-meter-modbus github.
https://github.com/Marvo2011/solar-meter-modbus/blob/main/SolaX%20Pocket%20Wifi%202.0%20API.md

There's a few API's they've listed on there.

**Comment by bjorn-lebel** [comment]
Hi, 
first off, great job with the plugin!
I think that this would be a great thing to implement!

What I miss, is to automate the inverter with actions in automation scripts.
Would like to have the option to set it to work mode "Manual - force charge" with the grid electricity price as trigger (for instance).

And to do that we need some "controls" on the plugin/device level.

(I have just installed the inverter so i havent checked what is possible and what is not with it yet)

**Comment by dakillerbee** [comment]
Hi,

i can provide you a little bit more input for setting inverter registers via local API.... For some reasons the Solax Modbus integration ist not working with my X1 Mini GEN3.

In my case i wanted to limit the power output to 60% of the max AC output. The register in my case was **17**.

![image](https://github.com/user-attachments/assets/0094c1b1-fd29-4568-bfc4-8a9a3d7af645)

e.g. max ac power 1100W output should be limited to 660 W  -> 60 %

_Request (Note that the double quotes are escaped since I do this from a windows dos console)_
`curl -d "optType=setReg&pwd=<Your_Inverter_SN>&data={\"num\":1,\"Data\":[{\"reg\":17,\"val\":\"60\"}]}" -X POST http://INVERTER_LOCAL_IP`

_Response:_
`Y:code,message:"success"}`

**Importat note:**
If you get the follwing response `0:code,message:"failed"}` then you have to send an unlock request (6868 is the default password):
`curl -d "optType=setReg&pwd=<Your_Inverter_SN>&data={\"num\":1,\"Data\":[{\"reg\":"0",\"val\":\"6868\"}]}" -X POST http://192.168.1.74`

For some reasons my inverter needs an unlock every morning after the startup.

And yeah thank you for this nice plugin

**Reply by anders-ae-dahl** [reply]
Thanks this works! I have a `X3-MIC/PRO-G2`, the Output Power Limit for that one is reg number 56. Used the inspector in Chrome to find the reg numbers while using the Solax Cloud web-interface.

**Reply by darmach** [reply]
@anders-ae-dahl would you advise / a quick howto on getting this with the inspector in Chrome?

**Reply by darmach** [reply]
@dakillerbee just a small question - why "60" in your curl example is passed in double quotes opposite to 17? Is that supposed to be passed as string?

I'm asking because in my case, after  sending this:

`curl -d "optType=setReg&pwd=**********&data={\"num\":1,\"Data\":[{\"reg\":28,\"val\":1}]}" -X POST http://192.168.1.10`

My API hang for about a minute or two, was unmanageable from the cloud. Gave me quite a scare!

My payload in web interface was:

```
optType setReg
num 1
deviceType 1
sn ***********
inverterSn *******************
tokenId afc65***********************1f
Data [{"reg":28,"val":0}]
```
Was my call correct?

**Reply by darmach** [reply]
@dakillerbee @anders-ae-dahl Could you guys take a look at my question above in spare moment? :)

**Comment by darmach** [comment]
@kclif9 Can we do some GET to get the current state of specific register? Thinking about creating a Homeassistant "restful switch" for self-use / feed priority on X3-hybrid-G4

**Comment by darmach** [comment]
Got it, it seem to be register 28 for X3 Hybrid G4:
![image](https://github.com/user-attachments/assets/4172891b-2fa7-4c41-9fac-ef1a6d37d2d7)

**Comment by darmach** [comment]
@kclif9 @anders-ae-dahl @dakillerbee Did any of you guys managed to get some progress with that?

**Comment by dakillerbee** [comment]
@darmach To add some more clarification regarding the double quotes...

The solax cloud api does not match the local api when you access the inverter locally... There seems to be some internal mapping/conversion done as far as I can say.

That is the reason why the local request uses double quote for the register value... I found it out by try and error 

With the following command you can read the registers:

`curl -d "?optType=ReadSetData&pwd=<Your_Inverter_SN> INVERTER_LOCAL_IP`

Please be aware that the position of the value in the output does NOT correpond with the register number.

In my case the _power limit factor_ value in the result is at the 13th position in the result but actually is set by using register 17.
